File: llm_harness/stats/data_sources/models_dev.py
"""models.dev stats source (API-only, failure-safe)."""

from __future__ import annotations

from datetime import UTC, datetime, timedelta
from typing import Any, TypedDict

import httpx

MODELS_DEV_URL = "https://models.dev/api.json"
LOOKBACK_DAYS = 365
REQUEST_TIMEOUT_SECONDS = 30.0


class ModelsDevOptions(TypedDict):
    """Reserved options for models.dev stats."""


class ModelsDevOutputPayload(TypedDict):
    """Flattened and ranked models.dev payload."""

    fetched_at_epoch_seconds: int | None
    status_code: int | None
    models: list[dict[str, Any]]


def _now_epoch_seconds() -> int:
    return int(datetime.now(UTC).timestamp())


def _iso_date_days_ago(days: int) -> str:
    return (datetime.now(UTC) - timedelta(days=days)).date().isoformat()


def _is_recent_date(iso_date: str | None, cutoff_iso_date: str) -> bool:
    if not iso_date:
        return False
    return iso_date >= cutoff_iso_date


def _as_finite(value: Any) -> float | None:
    try:
        numeric = float(value)
    except (TypeError, ValueError):
        return None
    if numeric != numeric or numeric in (float("inf"), float("-inf")):
        return None
    return numeric


def _fetch_models_dev() -> dict[str, Any]:
    with httpx.Client(timeout=REQUEST_TIMEOUT_SECONDS) as client:
        response = client.get(MODELS_DEV_URL)
    response.raise_for_status()
    payload = response.json()
    return {
        "fetched_at_epoch_seconds": _now_epoch_seconds(),
        "status_code": response.status_code,
        "payload": payload,
    }


def _flatten_models(payload: dict[str, Any]) -> list[dict[str, Any]]:
    rows: list[dict[str, Any]] = []
    for provider_id, provider in payload.items():
        if not isinstance(provider, dict):
            continue
        provider_name = provider.get("name") or provider_id
        models = provider.get("models") or {}
        if not isinstance(models, dict):
            continue
        for model_id, model in models.items():
            if not isinstance(model, dict):
                continue
            rows.append(
                {
                    "provider_id": provider_id,
                    "provider_name": provider_name,
                    "model_id": model.get("id") or model_id,
                    "model": model,
                }
            )
    return rows


def _rank_recent_models(models: list[dict[str, Any]], cutoff_iso_date: str) -> list[dict[str, Any]]:
    filtered = [row for row in models if _is_recent_date((row.get("model") or {}).get("release_date"), cutoff_iso_date)]
    filtered.sort(key=lambda row: (row.get("model") or {}).get("release_date") or "", reverse=True)
    filtered.sort(key=lambda row: _as_finite(((row.get("model") or {}).get("cost") or {}).get("output")) or float("inf"))
    return filtered


def get_models_dev_stats(
    _options: ModelsDevOptions | None = None,
) -> ModelsDevOutputPayload:
    """Fetch, flatten, and rank recent models from models.dev."""
    try:
        source_payload = _fetch_models_dev()
        cutoff_iso_date = _iso_date_days_ago(LOOKBACK_DAYS)
        return {
            "fetched_at_epoch_seconds": source_payload["fetched_at_epoch_seconds"],
            "status_code": source_payload["status_code"],
            "models": _rank_recent_models(_flatten_models(source_payload.get("payload") or {}), cutoff_iso_date),
        }
    except Exception:
        return {
            "fetched_at_epoch_seconds": None,
            "status_code": None,
            "models": [],
        }

File: llm-harness-js/src/stats/data-sources/modelsDev.ts
const MODELS_DEV_URL = "https://models.dev/api.json";
const LOOKBACK_DAYS = 365;
const REQUEST_TIMEOUT_MS = 30_000;

type NumberOrNull = number | null;

export type ModelRecord = {
  id?: string;
  name?: string;
  family?: string;
  release_date?: string;
  last_updated?: string;
  open_weights?: boolean;
  reasoning?: boolean;
  tool_call?: boolean;
  cost?: {
    input?: number;
    output?: number;
    cache_read?: number;
    cache_write?: number;
    output_audio?: number;
  };
  limit?: {
    context?: number;
    output?: number;
  };
  modalities?: {
    input?: string[];
    output?: string[];
  };
  [key: string]: unknown;
};

export type ProviderRecord = {
  id?: string;
  name?: string;
  api?: string;
  models?: Record<string, ModelRecord>;
  [key: string]: unknown;
};

export type ModelsDevPayload = Record<string, ProviderRecord>;

export type ModelsDevFlatModel = {
  provider_id: string;
  provider_name: string;
  model_id: string;
  model: ModelRecord;
};

type ModelsDevSourcePayload = {
  fetched_at_epoch_seconds: number | null;
  status_code: number | null;
  payload: ModelsDevPayload;
};

export type ModelsDevOutputPayload = {
  fetched_at_epoch_seconds: number | null;
  status_code: number | null;
  models: ModelsDevFlatModel[];
};

export type ModelsDevOptions = Record<string, never>;

function nowEpochSeconds(): number {
  return Math.floor(Date.now() / 1000);
}

function isoDateDaysAgo(days: number): string {
  return new Date(Date.now() - days * 24 * 60 * 60 * 1000)
    .toISOString()
    .slice(0, 10);
}

function isRecentDate(
  isoDate: string | undefined,
  cutoffIsoDate: string,
): boolean {
  if (!isoDate) {
    return false;
  }
  return isoDate >= cutoffIsoDate;
}

function asFinite(value: unknown): NumberOrNull {
  const numeric = Number(value);
  return Number.isFinite(numeric) ? numeric : null;
}

async function fetchModelsDev(): Promise<ModelsDevSourcePayload> {
  const controller = new AbortController();
  const timeout = setTimeout(() => controller.abort(), REQUEST_TIMEOUT_MS);
  const response = await fetch(MODELS_DEV_URL, { signal: controller.signal });
  clearTimeout(timeout);

  if (!response.ok) {
    throw new Error(`models.dev request failed: ${response.status}`);
  }

  const payload = (await response.json()) as ModelsDevPayload;
  const sourcePayload: ModelsDevSourcePayload = {
    fetched_at_epoch_seconds: nowEpochSeconds(),
    status_code: response.status,
    payload,
  };
  return sourcePayload;
}

function flattenModels(payload: ModelsDevPayload): ModelsDevFlatModel[] {
  const rows: ModelsDevFlatModel[] = [];
  for (const [providerId, provider] of Object.entries(payload)) {
    const providerName = provider.name ?? providerId;
    const models = provider.models ?? {};
    for (const [modelId, model] of Object.entries(models)) {
      rows.push({
        provider_id: providerId,
        provider_name: providerName,
        model_id: model.id ?? modelId,
        model,
      });
    }
  }
  return rows;
}

function rankRecentModels(
  models: ModelsDevFlatModel[],
  cutoffIsoDate: string,
): ModelsDevFlatModel[] {
  return models
    .filter((row) => isRecentDate(row.model.release_date, cutoffIsoDate))
    .sort((left, right) => {
      const leftOutputCost =
        asFinite(left.model.cost?.output) ?? Number.POSITIVE_INFINITY;
      const rightOutputCost =
        asFinite(right.model.cost?.output) ?? Number.POSITIVE_INFINITY;
      if (leftOutputCost !== rightOutputCost) {
        return leftOutputCost - rightOutputCost;
      }
      return (right.model.release_date ?? "").localeCompare(
        left.model.release_date ?? "",
      );
    });
}

export async function getModelsDevStats(
  _options: ModelsDevOptions = {},
): Promise<ModelsDevOutputPayload> {
  try {
    const sourcePayload = await fetchModelsDev();
    const cutoffIsoDate = isoDateDaysAgo(LOOKBACK_DAYS);
    return {
      fetched_at_epoch_seconds: sourcePayload.fetched_at_epoch_seconds,
      status_code: sourcePayload.status_code,
      models: rankRecentModels(
        flattenModels(sourcePayload.payload),
        cutoffIsoDate,
      ),
    };
  } catch {
    return {
      fetched_at_epoch_seconds: null,
      status_code: null,
      models: [],
    };
  }
}
